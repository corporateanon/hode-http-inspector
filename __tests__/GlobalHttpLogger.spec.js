const axios = require('axios');
const sleep = require('sleep-promise');
const GlobalHttpLogger = require('../lib/GlobalHttpLogger');
const { prepareSnapshot } = require('./utils');

const skipTick = () => sleep(0);

describe('GlobalHttpLogger', () => {
    it('should capture get requests', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({ onRequestEnd });
        logger.start();
        const res = await axios.get('http://127.0.0.1:13000/get?a=b');
        expect(res.data).toEqual('Hello World!');
        expect(res.status).toEqual(200);

        await skipTick();
        expect(
            prepareSnapshot(onRequestEnd.mock.calls[0][0])
        ).toMatchSnapshot();
    });

    it('should capture post requests', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({ onRequestEnd });
        logger.start();
        const res = await axios.post('http://127.0.0.1:13000/post', {
            foo: 'bar'
        });
        expect(res.data).toEqual({ hello: 'world' });
        expect(res.status).toEqual(200);

        await skipTick();
        expect(
            prepareSnapshot(onRequestEnd.mock.calls[0][0])
        ).toMatchSnapshot();
    });

    it('should capture get requests with error response', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({ onRequestEnd });
        logger.start();
        const res = await axios.get('http://127.0.0.1:13000/get/404', {
            validateStatus: () => true
        });
        expect(res.data).toEqual({ error: 'Not found' });
        expect(res.status).toEqual(404);

        await skipTick();
        expect(
            prepareSnapshot(onRequestEnd.mock.calls[0][0])
        ).toMatchSnapshot();
    });

    it('should capture get requests with network error', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({ onRequestEnd });
        logger.start();

        await expect(
            axios.get('http://never.existing.host.asdfgh/', {
                validateStatus: () => true
            })
        ).rejects.toThrow(
            'getaddrinfo ENOTFOUND never.existing.host.asdfgh never.existing.host.asdfgh:80'
        );

        await skipTick();
        expect(
            prepareSnapshot(onRequestEnd.mock.calls[0][0])
        ).toMatchSnapshot();
    });
});
