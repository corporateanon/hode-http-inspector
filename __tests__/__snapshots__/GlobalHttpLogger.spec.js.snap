// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalHttpLogger should capture get requests 1`] = `
Object {
  "error": null,
  "request": Object {
    "body": null,
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "host": "127.0.0.1:13000",
      "user-agent": "axios/0.19.0",
    },
    "method": "GET",
    "url": "http://127.0.0.1:13000/get?a=b",
  },
  "response": Object {
    "body": "Hello World!",
    "headers": Object {
      "connection": "close",
      "content-length": "12",
      "content-type": "text/html; charset=utf-8",
      "x-powered-by": "Express",
    },
    "status": 200,
  },
}
`;

exports[`GlobalHttpLogger should capture get requests with error response 1`] = `
Object {
  "error": null,
  "request": Object {
    "body": null,
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "host": "127.0.0.1:13000",
      "user-agent": "axios/0.19.0",
    },
    "method": "GET",
    "url": "http://127.0.0.1:13000/get/404",
  },
  "response": Object {
    "body": "{\\"error\\":\\"Not found\\"}",
    "headers": Object {
      "connection": "close",
      "content-length": "21",
      "content-type": "application/json; charset=utf-8",
      "x-powered-by": "Express",
    },
    "status": 404,
  },
}
`;

exports[`GlobalHttpLogger should capture get requests with network error 1`] = `
Object {
  "error": Object {
    "code": "ENOTFOUND",
    "message": "getaddrinfo ENOTFOUND never.existing.host.asdfgh never.existing.host.asdfgh:80",
  },
  "request": Object {
    "body": null,
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "host": "never.existing.host.asdfgh",
      "user-agent": "axios/0.19.0",
    },
    "method": "GET",
    "url": "http://never.existing.host.asdfgh/",
  },
  "response": null,
}
`;

exports[`GlobalHttpLogger should capture post requests 1`] = `
Object {
  "error": null,
  "request": Object {
    "body": "{\\"foo\\":\\"bar\\"}",
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "content-length": 13,
      "content-type": "application/json;charset=utf-8",
      "host": "127.0.0.1:13000",
      "user-agent": "axios/0.19.0",
    },
    "method": "POST",
    "url": "http://127.0.0.1:13000/post",
  },
  "response": Object {
    "body": "{\\"hello\\":\\"world\\"}",
    "headers": Object {
      "connection": "close",
      "content-length": "17",
      "content-type": "application/json; charset=utf-8",
      "x-powered-by": "Express",
    },
    "status": 200,
  },
}
`;
